File: CustomerTestTask.Data/DynamicLinqItems/FilterItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Dynamic.Core;

namespace CustomerTestTask.Data.DynamicLinqItems
{
    public class FilterItem<TEntity> where TEntity: class
    {
        private List<string> _operatorList = new List<string>{
            ">", "<", "==", "!=", ">=", "<=", "contains"
        };

        private string _customOperator;
        private string _name;

        public FilterItem(string name, string customOperator, string value)
        {
            Name = name;
            CustomOperator = customOperator;
            Value = value;
        }

        public string Name {
            get => _name; 
            set {            
                if (typeof(TEntity).GetProperty(value) == null) 
                {
                    throw new Exception($"Invalid property name {value}");
                }
                _name = value;
            }
        }
        public string CustomOperator {
            get => _customOperator; 
            set {
                if (_operatorList.Any(o => o == value) == false)
                {
                   throw new Exception($"Invalid operator {value}");
                }
                _customOperator = value; 
            }
        }

        public string Value {get; set;}
        
        public IQueryable<TEntity> AddFilter(IQueryable<TEntity> source)
        {
            if (CustomOperator == "contains") {
                return source.Where($"{Name}.Contains(@0)", Value);
            }

            return source.Where($"{Name} {CustomOperator} @0", Value);
        }
    }
}

File: CustomerTestTask.Client/src/app/items/api/filter.item.ts
export class FilterItem {
    public static readonly OPERATOR_LIST: string[] = [
        ">", "<", "==", "!=", ">=", "<=", "contains"
    ];

    private _name: string = '';
    private _customOperator: string = '';
    private _value: string = '';

    constructor(
        name: string, 
        customOperator: string, 
        value: string
    ) { 
        this.name = name;
        this.customOperator = customOperator;
        this.value = value;
    }

    public get name(): string {
        return this._name;
    }

    public set name(name: string) {        
        this._name = name;
    }

    public get customOperator(): string {
        return this._customOperator;
    }

    public set customOperator(customOperator: string) {
        if (FilterItem.OPERATOR_LIST.includes(customOperator) == false) {
            throw Error(`invalid operator - ${customOperator}`);
        }
        
        this._customOperator = customOperator;
    }

    public get value(): string {
        return this._value;
    }

    public set value(value: string) {        
        this._value = value;
    }

    public json(): string {
        return JSON.stringify({
            name: this.name,
            customOperator: this.customOperator,
            value: this.value
        });
    }

    public static fromJson(stringValue: string): FilterItem {
        const object: any = JSON.parse(stringValue);
        return new FilterItem(object['name'], object['customOperator'], object['value']); 
    }
}